src/thunderbots/software/geom/util.cpp

```diff
@@ -0,0 +1,939 @@
+                                        std::min(min1.y() * invertY, min2.y() * invertY));
+            else
+                minIntersection = Point(std::max(min1.x(), min2.x()),
+                                        std::max(min1.y() * invertY, min2.y() * invertY));
+
+            Point maxIntersection;
+            if (isDescending)
+                maxIntersection = Point(std::min(max1.x(), max2.x()),
+                                        std::max(max1.y() * invertY, max2.y() * invertY));
+            else
+                maxIntersection = Point(std::min(max1.x(), max2.x()),
+                                        std::min(max1.y() * invertY, max2.y() * invertY));
+
+            bool intersect =
+                minIntersection.x() <= maxIntersection.x() &&
+                ((!isDescending && minIntersection.y() <= maxIntersection.y()) ||
+                 (isDescending && minIntersection.y() >= maxIntersection.y()));
+
+            if (intersect)
+            {
+                return std::vector<Point>{minIntersection, maxIntersection};
+            }
+            else
+                return std::vector<Point>();
+        }
+        else
+            return std::vector<Point>();
+    }
+
+    return std::vector<Point>{a.start +
+                              (a.start - b.start).cross(b.end - b.start) /
+                                  (b.end - b.start).cross(a.end - a.start) *
+                                  (a.end - a.start)};
+}
+
+// shamelessly copy-pasted from RoboJackets
+std::optional<Point> lineIntersection(const Vector &a, const Vector &b, const Vector &c,
+                                      const Vector &d)
+{
+    Seg line1(a, b), line2(c, d);
+    double x1 = line1.start.x();
+    double y1 = line1.start.y();
+    double x2 = line1.end.x();
+    double y2 = line1.end.y();
+    double x3 = line2.start.x();
+    double y3 = line2.start.y();
+    double x4 = line2.end.x();
+    double y4 = line2.end.y();
+
+    double denom = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4);
+    if (denom == 0)
+    {
+        // log the parallel lines when we actually implement logging?
+        return std::nullopt;
+    }
+
+    double deta = x1 * y2 - y1 * x2;
+    double detb = x3 * y4 - y3 * x4;
+
+    Point intersection;
+
+    intersection.set((deta * (x3 - x4) - (x1 - x2) * detb) / denom,
+                     (deta * (y3 - y4) - (y1 - y2) * detb) / denom);
+
+    return std::make_optional(intersection);
+}
+
+
+Vector reflect(const Vector &v, const Vector &n)
+{
+    if (n.len() < EPS)
+    {
+        return v;
+    }
+    Vector normal = n.norm();
+    return v - 2 * v.dot(normal) * normal;
+}
+
+Vector reflect(const Vector &a, const Vector &b, const Vector &p)
+{
+    // Make a as origin.
+    // Rotate by 90 degrees, does not matter which direction?
+    Vector n = (b - a).rotate(Angle::quarter());
+    return a + reflect(p - a, n);
+}
+
+// ported code
```

Comment:
Are all the following `// ported code` comments needed? Can be removed otherwise.

Reply:
removed